File: authentication-server/src/main/java/com/andrew/authenticationserver/service/additional/JwtTokenService.java
package com.andrew.authenticationserver.service.additional;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Component
public class JwtTokenService {
    @Value("${jwt.token.secret.access}")
    private String jwtAccessSecret;
    @Value("${jwt.token.duration}") long durationMinute;

    public String generateAccessToken(UserDetails user) {
        Map<String, Object> claims = new HashMap<>();
        List<String> rolesList = user.getAuthorities().stream()
                .map(GrantedAuthority::getAuthority)
                .collect(Collectors.toList());
        claims.put("roles", rolesList);
        LocalDateTime nowDate = LocalDateTime.now();
        Date expireDate = Date.from(nowDate.plusMinutes(durationMinute).atZone(ZoneId.systemDefault()).toInstant());

        return Jwts.builder()
                .setClaims(claims)
                .setSubject(user.getUsername())
                .setExpiration(expireDate)
                .signWith(SignatureAlgorithm.HS256, jwtAccessSecret)
                .compact();
    }

    public boolean validateAccessToken(String token) {
        return validateToken(token, jwtAccessSecret);
    }

    public String getUsername(String token) {
        return getAllClaimsFromToken(token).getSubject();
    }

    public List<String> getRoles(String token) {
        return getAllClaimsFromToken(token).get("roles", List.class);
    }

    private boolean validateToken(String token, String secret) {
        try {
            Jwts.parser()
                    .setSigningKey(secret)
                    .parseClaimsJws(token);
            return true;
        } catch (Exception e) {
            return false;
        }
    }
    private Claims getAllClaimsFromToken(String token) {
        return Jwts.parser()
                .setSigningKey(jwtAccessSecret)
                .parseClaimsJws(token)
                .getBody();
    }
}

File: auth-server/src/main/java/com/andrew/authserver/service/add_serv/JwtTokenService.kt
package com.andrew.authserver.service.add_serv

import com.andrew.authserver.entity.auth.Role
import com.andrew.authserver.entity.auth.User
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import javax.crypto.SecretKey
import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import lombok.NonNull

import java.security.Key

import java.time.LocalDateTime
import java.time.ZoneId
import java.util.Date


@Service
open class JwtTokenService (
        @Value("\${jwt.token.secret.access}") jwtAccessSecret: String,
        @Value("\${jwt.token.duration}") durationMinute: Long
        ) {

        private val jwtAccessSecret: SecretKey = Keys.hmacShaKeyFor(Decoders.BASE64.decode(jwtAccessSecret));
        private val duration: Long = durationMinute;

        //generate of access token
        fun generateAccessToken(user: User): String{
                val nowDate = LocalDateTime.now();
                val expireInstant = nowDate.plusMinutes(duration).atZone(ZoneId.systemDefault()).toInstant();
                val expireDate = Date.from(expireInstant);

                return Jwts.builder()
                        .setSubject(user.getUsername())
                        .setExpiration(expireDate)
                        .signWith(jwtAccessSecret)
                        .claim("roles", user.getRoles())
                        .claim("email", user.getEmail())
                        .compact();
        }

        fun validateAccessToken(@NonNull token: String): Boolean {
                return validateToken(token, jwtAccessSecret);
        }

        fun getUsername(token: String): String{
                return getAllClaimsFromToken(token).subject;
        }

        fun getRoles(token: String): List<Role> {
                val claims = getAllClaimsFromToken(token)

                val roles = claims.get("roles", List::class.java) as? List<*> ?: return emptyList()

                return roles.mapNotNull { it as? Role }
        }

        private fun validateToken(@NonNull token: String, @NonNull secret: Key): Boolean{
                return try{
                        Jwts.parserBuilder()
                                .setSigningKey(jwtAccessSecret)
                                .build()
                                .parseClaimsJws(token);
                        true;
                } catch (e: Exception){
                        false;
                }
        }

        private fun getAllClaimsFromToken(token: String): Claims{
                return Jwts.parser()
                        .setSigningKey(jwtAccessSecret)
                        .parseClaimsJws(token)
                        .body;
        }
}